core/src/main/scala/kafka/server/ReplicationQuotaManager.scala

```diff
@@ -0,0 +1,192 @@
+  * @param metrics         The Metrics instance
+  * @param replicationType The name / key for this quota manager, typically leader or follower
+  * @param time            Time object to use
+  */
+class ReplicationQuotaManager(val config: ReplicationQuotaManagerConfig,
+                              private val metrics: Metrics,
+                              private val replicationType: String,
+                              private val time: Time) extends Logging with ReplicaQuota {
+  private val lock = new ReentrantReadWriteLock()
+  private val throttledPartitions = new ConcurrentHashMap[String, Seq[Int]]()
+  private var quota: Quota = null
+  private val sensorAccess = new SensorAccess
+  private val rateMetricName = metrics.metricName("byte-rate", replicationType, s"Tracking byte-rate for $replicationType")
+
+  /**
+    * Update the quota
+    *
+    * @param quota
+    */
+  def updateQuota(quota: Quota) = {
+    inWriteLock(lock) {
+      this.quota = quota
+      if (metrics.metrics.containsKey(rateMetricName))
+        metrics.metrics.get(rateMetricName).config(getQuotaMetricConfig(quota))
+    }
+  }
+
+  /**
+    * Check if the quota is currently exceeded
+    *
+    * @return
+    */
+  override def isQuotaExceeded(): Boolean = {
+    try {
+      sensor.checkQuotas()
+    } catch {
+      case qve: QuotaViolationException =>
+        if(logger.isTraceEnabled)
+          logger.trace("%s: Quota violated for sensor (%s), metric: (%s), metric-value: (%f), bound: (%f)".format(replicationType, sensor.name(), qve.metricName, qve.value, qve.bound))
+        return true
+    }
+    false
+  }
+
+  /**
+    * Is the passed partition throttled by this ReplicationQuotaManager
+    *
+    * @param topicPartition the partition to check
+    * @return
+    */
+  override def isThrottled(topicPartition: TopicAndPartition): Boolean = {
+    val partitions = throttledPartitions.get(topicPartition.topic)
+    if (partitions != null)
+      partitions.contains(topicPartition.partition) || partitions == allReplicas
+    else false
+  }
+
+  /**
+    * Add the passed value to the throttled rate. This method ignores the quota with
+    * the value being added to the rate even if the quota is exceeded
+    *
+    * @param value
+    */
+  def record(value: Long) = {
+    try {
+      sensor.record(value)
+    } catch {
+      case qve: QuotaViolationException =>
+        if(logger.isTraceEnabled)
+          logger.trace("Record: Quota violated, but ignored, for sensor (%s), metric: (%s), metric-value: (%f), bound: (%f), recordedValue (%s)".format(sensor.name(), qve.metricName, qve.value, qve.bound, value))
+    }
+  }
+
+  /**
+    * Update the set of throttled partitions for this QuotaManager. The partitions passed, for
+    * any single topic, will replace any previous
+    *
+    * @param topic
+    * @param partitions the set of throttled partitions
+    * @return
+    */
+  def markThrottled(topic: String, partitions: Seq[Int]): Unit = {
```

Comment:
OK - thanks for the heads up - have changed tidied up where this was wrong elsewhere too.